pyfr/readers/cgns.py

```
@@ -0,0 +1,435 @@
+
+        # Read nodes
+        self.nodepts = self._read_nodepts(zone)
+
+        # Read bc
+        bc = self._read_bc(zone)
+
+        # Read elements
+        self.elenodes = elenodes = {}
+        self.pents = pents = {}
+
+        # Construct elenodes and physical entity
+        for idx in range(cgns.nsections(zone)):
+            elerng, elenode = self._read_element(zone, idx)
+
+            for bcname, bcrng in bc.items():
+                if elerng[0] >= bcrng[0] and elerng[1] <= bcrng[1]:
+                    name = bcname
+                    break
+                else:
+                    name = 'fluid'
+
+            pent = pents.setdefault(name, idx)
+
+            elenodes.update({(k, pent): v for k, v in elenode.items()})
+
+    def _read_nodepts(self, zone):
+        nnode = zone['size'][0]
+        nodepts = np.zeros((3, nnode))
+
+        self._cgns.coord_read(zone, 'CoordinateX', nodepts[0])
+        self._cgns.coord_read(zone, 'CoordinateY', nodepts[1])
+        self._cgns.coord_read(zone, 'CoordinateZ', nodepts[2])
+
+        return nodepts
+
+    def _read_bc(self, zone):
+        nbc = self._cgns.nbocos(zone)
+        bc = {}
+
+        for idx_bc in range(nbc):
+            boco = self._cgns.boco_read(zone, idx_bc)
+            name = boco['name'].lower()
+            bc[name] = boco['range']
+
+        return bc
+
+    def _read_element(self, zone, idx):
+        s = self._cgns.section_read(zone, idx)
+
+        elerng = s['range']
+        conn = np.zeros(s['dim'], dtype=np.int32)
+        self._cgns.elements_read(s, conn)
+
+        cgns_type = s['etype']
+        petype = self.cgns_map[cgns_type][0]
+
+        elenode = {}
+
+        if petype == 'mixed':
+            i = 0
+            mele = defaultdict(list)
+
+            while i < s['dim']:
+                try:
+                    cgns_type = conn[i]
+                    petype, spts = self.cgns_map[cgns_type]
+                except KeyError as err:
```

Just let this raise.